proxy/src/main/java/com/wavefront/agent/preprocessor/SpanWhitelistAnnotationTransformer.java

```diff
@@ -0,0 +1,78 @@
+package com.wavefront.agent.preprocessor;
+
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import wavefront.report.Annotation;
+import wavefront.report.Span;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Only allow span annotations that match the whitelist.
+ *
+ * @author vasily@wavefront.com
+ */
+public class SpanWhitelistAnnotationTransformer implements Function<Span, Span> {
+  private static final Set<String> SYSTEM_TAGS = ImmutableSet.of("service", "application",
+      "cluster", "shard");
+
+  private final Map<String, Pattern> whitelistedKeys;
+  private final PreprocessorRuleMetrics ruleMetrics;
+
+  SpanWhitelistAnnotationTransformer(final Map<String, String> keys,
+                                     final PreprocessorRuleMetrics ruleMetrics) {
+    this.whitelistedKeys = new HashMap<>(keys.size() + SYSTEM_TAGS.size());
+    SYSTEM_TAGS.forEach(x -> whitelistedKeys.put(x, null));
+    keys.forEach((k, v) -> whitelistedKeys.put(k, v == null ? null : Pattern.compile(v)));
+    Preconditions.checkNotNull(ruleMetrics, "PreprocessorRuleMetrics can't be null");
+    this.ruleMetrics = ruleMetrics;
+  }
+
+  @Nullable
+  @Override
+  public Span apply(@Nullable Span span) {
+    if (span == null) return null;
+    long startNanos = ruleMetrics.ruleStart();
+    List<Annotation> annotations = span.getAnnotations().stream().
+        filter(x -> {
+          if (!whitelistedKeys.containsKey(x.getKey())) return false;
```

Comment:
technically you can express this as two filters for slightly easier readability

Reply:
@panghy fixed, this should be more readable